### PR TITLE
Add Ignition support

### PR DESCRIPTION
**Pull request by hectorj2f**
We want to add support for Ignition as an additional bootstrap mechanism for CoreOS setups.

- [x] Add a `use-ignition` argument to the cli
- [x] Test the template generation
- [x]  Test on a real AWS cluster

### CODE REVIEW

**Comment by hectorj2f** [comment]
Ideally, we could try to gather the CoreOS version of the AMI to verify if it supports ignition.

File: cli/create.go [hunk 1]
```diff
@@ -37,7 +39,7 @@ func registerCreateFlags(flagset *pflag.FlagSet) {
 	flagset.String("etcd-discovery-url", "", "etcd discovery url for a secondary swarm is connecting to")
 	flagset.String("template-dir", "templates", "directory to use for reading templates (see template-init command)")
 
-	awsEuWest1CoreOS := "ami-5f2f5528" // CoreOS Stable 681.2.0 (HVM eu-west-1)
+	 // CoreOS Stable 681.2.0 (HVM eu-west-1)
```
Comment (zeisss) [inline cli/create.go]: put that comment please to the const

Comment (hectorj2f) [inline cli/create.go]: Yes

File: cli/create.go [hunk 2]
```diff
@@ -85,6 +89,10 @@ func runCreate(args []string) (exit int) {
 		return exitError("couldn't create swarm: --image must be provided")
 	}
 
+	if flags.ImageURI == awsEuWest1CoreOS && flags.UseIgnition {
+		return exitError(fmt.Sprintf("couldn't create swarm: --use-ignition requires a recent CoreOS AMI than '%s'", awsEuWest1CoreOS))
```
Comment (zeisss) [inline cli/create.go]: I don't get this check. What do we want to achieve here? also I think there is a word missing in the error text.

File: default-templates/primary-ignition.tmpl [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            Environment="ETCD_DISCOVERY={{.DiscoveryUrl}}"
+            Environment="ETCD_ADVERTISE_CLIENT_URLS=http://$private_ipv4:2379"
+            Environment="ETCD_INITIAL_ADVERTISE_PEER_URLS=http://$private_ipv4:2380"
+            Environment="ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379"
+            Environment="ETCD_LISTEN_PEER_URLS=http://$private_ipv4:2380,http://127.0.0.1:2380"
+            Environment="ETCD_ELECTION_TIMEOUT=5000"
+    - name: fleet.service
+      enable: true
+      dropins:
+        - name: 30-fleet-giantswarm.conf
```
Comment (zeisss) [inline default-templates/primary-ignition.tmpl]: nothing gs specific in the public repo

File: kocho.yml.example [hunk 4]
```diff
@@ -28,6 +28,8 @@ certificate: <insert ssl certificate arn here>
 # yochu-k8s-version: <k8s version>
 # yochu-rkt-version: <rkt version>
 
+# Use ignition
```
Comment (zeisss) [inline kocho.yml.example]: do we want to add a link to Ignition here? Just to give readers a hint what this is about?

File: cli/create.go (hunk 2 above)
Comment (hectorj2f) [inline cli/create.go]: ignition doesn't work with our default CoreOS version cause it is old, you need to use newer ones

File: default-templates/primary-ignition.tmpl (hunk 3 above)
Comment (hectorj2f) [inline default-templates/primary-ignition.tmpl]: ack

File: kocho.yml.example (hunk 4 above)
Comment (hectorj2f) [inline kocho.yml.example]: Yes, and documentation too.

File: cli/create.go (hunk 2 above)
Comment (zeisss) [inline cli/create.go]: You probably want `if flags.UseIgnition && !cmdCreate.Flags.Changed("image") {` then which feels more expressive to me for the intent.

> --use-ignition requires a recent CoreOS AMI than

How about
> ...--use-ignition requires a more recent CoreOS AMI than ...

**Comment by zeisss** [comment]
only minor comments ;) nothing major

File: cli/create.go (hunk 2 above)
Comment (hectorj2f) [inline cli/create.go]: done, but the proposal of using `cmdCreate` creates a `initialization loop`

File: default-templates/primary-ignition.tmpl (hunk 3 above)
Comment (hectorj2f) [inline default-templates/primary-ignition.tmpl]: done

**Comment by hectorj2f** [comment]
@zeisss Do you need someone else review to give me a LGTM ? I mostly addressed all your comments.

**Comment by hectorj2f** [comment]
ping @zeisss for an answer

**Comment by zeisss** [comment]
LGTM, if it works for you, feel free to merge.

**Comment by hectorj2f** [comment]
@zeisss It works but I found a weird behavior with the etcd cluster creation. I need to test it more

**Comment by hectorj2f** [comment]
@zeisss It works and the issue of the etcd out of the box creation. It also happens with cloudconfig setup so it should be a different thing and totally independent of this PR.

File: default-templates/primary-ignition.tmpl [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      enable: false
+      mask: true
+    - name: etcd2.service
+      enable: true
+      dropins:
+       - name: 30-etcd2-kocho.conf
+         contents: |
+           [Unit]
+           Requires=coreos-metadata.service
+           After=coreos-metadata.service
```
Comment (hectorj2f) [inline default-templates/primary-ignition.tmpl]: Variable substitution doesn't work with ignition like was working for cloudconfig. To get the value of this environment variables we need to make use of coreos-metadata service as done in here.

File: default-templates/primary-ignition.tmpl [hunk 6]
```diff
@@ -10,22 +10,29 @@ systemd:
...
+               --advertise-client-urls=http://${COREOS_EC2_IPV4_LOCAL}:2379 \
+               --initial-advertise-peer-urls=http://${COREOS_EC2_IPV4_LOCAL}:2380 \
+               --listen-client-urls=http://0.0.0.0:2379 \
+               --listen-peer-urls="http://${COREOS_EC2_IPV4_LOCAL}:2380,http://127.0.0.1:2380" \
+               --election-timeout=5000
     - name: fleet.service
       enable: true
       dropins:
-        - name: 10-fleet-kocho.conf
+        - name: 30-fleet-kocho.conf
```
Comment (zeisss) [inline default-templates/primary-ignition.tmpl]: I think the point of having numbers in here is that we can define an order with them and let one overwrite the other. So we should _not_ have the same numbers twice :D

File: default-templates/primary-ignition.tmpl [hunk 7]
```diff
@@ -10,22 +10,29 @@ systemd:
...
+           [Unit]
+           Requires=coreos-metadata.service
+           After=coreos-metadata.service
+
+           [Service]
+           EnvironmentFile=/run/metadata/coreos
+           ExecStart=
+           ExecStart=/usr/bin/etcd2 \
+               --discovery={{.DiscoveryUrl}} \
+               --advertise-client-urls=http://${COREOS_EC2_IPV4_LOCAL}:2379 \
```
Comment (zeisss) [inline default-templates/primary-ignition.tmpl]: Is this from Ignition?

Comment (hectorj2f) [inline default-templates/primary-ignition.tmpl]: It comes with the latest versions and it is normally used by ignition in such a cases

File: default-templates/primary-ignition.tmpl (hunk 6 above)
Comment (hectorj2f) [inline default-templates/primary-ignition.tmpl]: That is true but my intention is different. There are other configuration files coreos default (10-....), you want your configuration file to be the last one to be applied, so the numbers make sense to me here.

**Comment by zeisss** [comment]
still lgtm ;)

**Pull request merged**